File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksAdapter.java
package com.example.android.architecture.blueprints.todoapp.tasks;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.CheckBox;
import android.widget.TextView;

import com.example.android.architecture.blueprints.todoapp.R;
import com.example.android.architecture.blueprints.todoapp.data.Task;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

import static com.google.common.base.Preconditions.checkNotNull;

public class TasksAdapter extends BaseAdapter {

    private PublishSubject<Task> mTaskClickObservable = PublishSubject.create();
    private PublishSubject<Task> mTaskToggleObservable = PublishSubject.create();
    private List<Task> mTasks;

    public TasksAdapter(List<Task> tasks) {
        setList(tasks);
    }

    public void replaceData(List<Task> tasks) {
        setList(tasks);
        notifyDataSetChanged();
    }

    Observable<Task> getTaskClickObservable() {
        return mTaskClickObservable;
    }

    Observable<Task> getTaskToggleObservable() {
        return mTaskToggleObservable;
    }

    private void setList(List<Task> tasks) {
        this.mTasks = checkNotNull(tasks);
    }

    @Override
    public int getCount() {
        return mTasks.size();
    }

    @Override
    public Task getItem(int position) {
        return mTasks.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View view, ViewGroup viewGroup) {
        View rowView = view;
        if (rowView == null) {
            LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
            rowView = inflater.inflate(R.layout.task_item, viewGroup, false);
        }

        final Task task = getItem(position);

        TextView titleTV = (TextView) rowView.findViewById(R.id.title);
        titleTV.setText(task.getTitleForList());

        CheckBox completeCB = (CheckBox) rowView.findViewById(R.id.complete);

        // Active/completed task UI
        completeCB.setChecked(task.isCompleted());
        if (task.isCompleted()) {
            rowView.setBackgroundDrawable(viewGroup.getContext()
                    .getResources()
                    .getDrawable(R.drawable.list_completed_touch_feedback));
        } else {
            rowView.setBackgroundDrawable(
                    viewGroup.getContext().getResources().getDrawable(R.drawable.touch_feedback));
        }

        completeCB.setOnClickListener(ignored -> mTaskToggleObservable.onNext(task));

        rowView.setOnClickListener(ignored -> mTaskClickObservable.onNext(task));

        return rowView;
    }
}

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/tasks/TasksAdapter.kt
package com.example.android.architecture.blueprints.todoapp.tasks

import android.support.v4.content.ContextCompat
import android.support.v4.view.ViewCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.CheckBox
import android.widget.TextView
import com.example.android.architecture.blueprints.todoapp.R
import com.example.android.architecture.blueprints.todoapp.data.Task
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject

class TasksAdapter(tasks: List<Task>) : BaseAdapter() {
  private val taskClickSubject = PublishSubject.create<Task>()
  private val mTaskToggleObservable = PublishSubject.create<Task>()
  private lateinit var tasks: List<Task>

  val taskClickObservable: Observable<Task>
    get() = taskClickSubject

  val taskToggleObservable: Observable<Task>
    get() = mTaskToggleObservable

  init {
    setList(tasks)
  }

  fun replaceData(tasks: List<Task>) {
    setList(tasks)
    notifyDataSetChanged()
  }

  private fun setList(tasks: List<Task>) {
    this.tasks = tasks
  }

  override fun getCount(): Int = tasks.size

  override fun getItem(position: Int): Task = tasks[position]

  override fun getItemId(position: Int): Long = position.toLong()

  override fun getView(position: Int, view: View?, viewGroup: ViewGroup): View {
    val rowView: View = view ?:
        LayoutInflater.from(viewGroup.context).inflate(R.layout.task_item, viewGroup, false)

    val task = getItem(position)

    rowView.findViewById<TextView>(R.id.title).text = task.titleForList

    val completeCB = rowView.findViewById<CheckBox>(R.id.complete)

    // Active/completed task UI
    completeCB.isChecked = task.completed
    if (task.completed) {
      ViewCompat.setBackground(
          rowView,
          ContextCompat.getDrawable(viewGroup.context, R.drawable.list_completed_touch_feedback))
    } else {
      ViewCompat.setBackground(
          rowView,
          ContextCompat.getDrawable(viewGroup.context, R.drawable.touch_feedback))
    }

    completeCB.setOnClickListener { mTaskToggleObservable.onNext(task) }

    rowView.setOnClickListener { taskClickSubject.onNext(task) }

    return rowView
  }
}
